### PR TITLE
Clean up nvtx macros

### PR DESCRIPTION
**Pull request by PointKernel**
## Description
This PR includes several cleanups for the cudf nvtx wrappers:

- Removed the unused `NVTX3_FUNC_RANGE` macro
- Fixed a typo in the doc
- Added an example in the `cudf::thread_range` doc
- Updated the `NVTX` section in the developer guide doc

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by PointKernel** [comment]
Users cannot specify the function name with the current `CUDF_FUNC_RANGE`: 
Screenshot from 2024-02-13 11-51-04
This PR allows custom naming, i.e. `hash_join::inner_join`:
Screenshot from 2024-02-13 11-49-56

I've also considered using `__PRETTY_FUNCTION__` macro to replace `__func__` but realized two issues by doing so:

1. Unlike `__func__`, `__PRETTY_FUNCTION__` is not a standard macro
2. It returns the complete syntax of class member functions, which makes the result difficult to read, e.g.
Screenshot from 2024-02-13 11-53-02

**Comment by davidwendt** [comment]
You could always just call `nvtxRangePushA(custom-name)' and 'nvtxRangePop()`

File: cpp/include/cudf/detail/nvtx/ranges.hpp [hunk 1]
```diff
@@ -37,15 +37,26 @@ using thread_range = ::nvtx3::domain_thread_range<libcudf_domain>;
...
- * Uses the name of the immediately enclosing function returned by `__func__` to
- * name the range.
+ * Uses the given name or the name of the immediately enclosing function returned
+ * by `__func__` to name the range.
  *
  * Example:
  * ```
  * void some_function(){
  *    CUDF_FUNC_RANGE();
+ *    CUDF_FUNC_RANGE("custom_name"); // Or use a custom name
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/nvtx/ranges.hpp:47]: This macro should not exist. The _only_ reason for `CUDF_FUNC_RANGE` is for when you want to use the name of the function as the name of the range.

If you want to provide a custom name, then use a range object directly:
```
cudf::thread_range rng{"custom_name"};
```

**Comment by jrhemstad** [comment]
Also, this reminds me that the vendored version of nvtx in cudf is wildly out of date and should be removed. See https://github.com/rapidsai/cudf/issues/6476

File: cpp/include/cudf/detail/nvtx/ranges.hpp (hunk 1 above)
Comment (PointKernel) [reply cpp/include/cudf/detail/nvtx/ranges.hpp:47]: Ah, good to know.

I will change this PR for several minor cleanups and doc improvements then. https://github.com/rapidsai/cudf/issues/6476 should be addressed separately.

Comment (bdice) [reply cpp/include/cudf/detail/nvtx/ranges.hpp]: Thanks @jrhemstad -- I had not come across `cudf::thread_range` before.

@PointKernel Can you update the [Developer Guide's NVTX Ranges section](https://github.com/rapidsai/cudf/blob/branch-24.04/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#nvtx-ranges) to describe exactly this:

- Use `CUDF_FUNC_RANGE` if you want to use the name of the function as the name of the NVTX range
- Use `cudf::thread_range rng{"custom_name"};` to provide a custom name for the current scope's NVTX range

Comment (PointKernel) [reply cpp/include/cudf/detail/nvtx/ranges.hpp]: 2c45d7a24c506e8e9364a91a2886d5b81aabc8d0 Done.

Comment (karthikeyann) [reply cpp/include/cudf/detail/nvtx/ranges.hpp]: Related discussion with Jake
https://github.com/rapidsai/cudf/pull/11864#issuecomment-1271514499

**Review by karthikeyann — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Pull request merged**